File: stego/bitio.py
"""Bit packing, a JS-compatible PRNG, deterministic interleaving, and CRC-16.

The PRNG is a byte-for-byte port of the browser's mulberry32 (`smb`) and Fisher-Yates
(`sperm`), so a payload embedded by Python in a WAV is decodable by the JS app and vice
versa. Everything operates on uint32 bit-patterns to match JS's 32-bit integer semantics.
"""
import numpy as np

_MASK = 0xFFFFFFFF

def _u32(x): return x & _MASK

def mulberry32(seed):
    """Port of JS  function smb(a){...}  — returns a float-in-[0,1) generator."""
    a = _u32(seed)
    def rnd():
        nonlocal a
        a = _u32(a + 0x6D2B79F5)
        t = _u32(_u32(a ^ (a >> 15)) * _u32(1 | a))
        t = _u32(_u32(t + _u32(_u32(t ^ (t >> 7)) * _u32(61 | t))) ^ t)
        return _u32(t ^ (t >> 14)) / 4294967296.0
    return rnd

def perm(n, seed):
    """Port of JS sperm(n,seed): seeded Fisher-Yates permutation of [0..n)."""
    r = mulberry32(seed)
    a = list(range(n))
    for i in range(n - 1, 0, -1):
        j = int(r() * (i + 1))
        a[i], a[j] = a[j], a[i]
    return np.asarray(a, dtype=np.int64)

# ---- bits <-> bytes (MSB first within each byte; matches JS sbits/sunbits) ----
def bytes_to_bits(data):
    return np.unpackbits(np.frombuffer(data, dtype=np.uint8))

def bits_to_bytes(bits):
    return np.packbits(np.asarray(bits, dtype=np.uint8)).tobytes()

# ---- CRC-16/CCITT-FALSE (per-symbol integrity; cheap, turns corruption into erasure) ----
def crc16(data):
    crc = 0xFFFF
    for b in data:
        crc ^= b << 8
        for _ in range(8):
            crc = ((crc << 1) ^ 0x1021) & 0xFFFF if (crc & 0x8000) else (crc << 1) & 0xFFFF
    return crc

File: web/src/stego/bitio.js
// Bit packing, mulberry32 PRNG, CRC-16 and CRC-32 — byte-identical to the Python `stego` package
// so a stego WAV produced by Python is decodable in the browser, and vice versa.

// `& 0xFFFFFFFF` is a TRAP in JS — bitwise ops are 32-bit *signed*, so `x & 0xFFFFFFFF`
// is signed-equivalent to `x` (e.g. `0xFFFFFFFF & 0xFFFFFFFF == -1`). Use `>>> 0` instead.
export const u32 = (x) => x >>> 0;

// mulberry32 — same algorithm as stego/bitio.py:mulberry32
export function mulberry32(seed) {
  let a = u32(seed);
  return function () {
    a = u32(a + 0x6D2B79F5);
    let t = u32(Math.imul(a ^ (a >>> 15), 1 | a));
    t = u32(u32(t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t);
    return u32(t ^ (t >>> 14)) / 4294967296;
  };
}

// Seeded Fisher–Yates — same algorithm as stego/bitio.py:perm
export function perm(n, seed) {
  const r = mulberry32(seed);
  const a = new Int32Array(n);
  for (let i = 0; i < n; i++) a[i] = i;
  for (let i = n - 1; i > 0; i--) {
    const j = Math.floor(r() * (i + 1));
    const t = a[i]; a[i] = a[j]; a[j] = t;
  }
  return a;
}

// bits <-> bytes (MSB first within each byte; matches numpy unpackbits/packbits)
export function bytesToBits(data) {
  const u8 = data instanceof Uint8Array ? data : new Uint8Array(data);
  const out = new Uint8Array(u8.length * 8);
  for (let i = 0; i < u8.length; i++) {
    const b = u8[i];
    for (let k = 0; k < 8; k++) out[i * 8 + k] = (b >> (7 - k)) & 1;
  }
  return out;
}
export function bitsToBytes(bits) {
  const out = new Uint8Array(bits.length >> 3);
  for (let i = 0; i < out.length; i++) {
    let v = 0;
    for (let k = 0; k < 8; k++) v = (v << 1) | bits[i * 8 + k];
    out[i] = v;
  }
  return out;
}

// CRC-16/CCITT-FALSE (poly 0x1021, init 0xFFFF) — matches stego/bitio.py:crc16
export function crc16(data) {
  let c = 0xFFFF;
  for (let i = 0; i < data.length; i++) {
    c ^= data[i] << 8;
    for (let k = 0; k < 8; k++) c = (c & 0x8000) ? (((c << 1) ^ 0x1021) & 0xFFFF) : ((c << 1) & 0xFFFF);
  }
  return c;
}

// CRC-32 (IEEE, poly 0xEDB88320 reflected) — matches Python zlib.crc32
const _CRC32_TABLE = (() => {
  const t = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) c = (c & 1) ? (0xEDB88320 ^ (c >>> 1)) : (c >>> 1);
    t[n] = c >>> 0;
  }
  return t;
})();
export function crc32(data) {
  let c = 0xFFFFFFFF;
  for (let i = 0; i < data.length; i++) c = _CRC32_TABLE[(c ^ data[i]) & 0xFF] ^ (c >>> 8);
  return (c ^ 0xFFFFFFFF) >>> 0;
}
